### ISSUE TITLE
Autocompletion broken in custom commands 

### ISSUE THREAD

**Issue by dyskos** [nushell/nushell#14238]
### Describe the bug

After upgrading from 0.98.0 -> 0.99.1, autocomplete doesn't work correctly for custom commands anymore. After typing the command and then pressing tab, the autocompletion for the first argument will show the list of local files as expected, but arguments 2 and further will simply show `NO RECORDS FOUND`. This is a regression compared to the behaviour of 0.98.0 as custom commands such as `lx hare run mai<tab>` could be reliably used to autocomplete for files and folders in the local directory.

This is the case on each system I've tested, both Windows 10 and Linux (Arch).

### How to reproduce

- Install Nushell 0.99.1 with a fresh default configuration
- Create a custom command in `config.nu` which takes multiple arguments or ...rest args
e.g.
```nu
def list [...args] {
    $args
    | each {
        open $args
    }
}
```
- Attempt to use the command on several files,
```
~> list main.c test.<tab>
```
- The second and further arguments will not autocomplete with local files/folders, and only show `NO RECORDS FOUND`

### Expected behavior

I expected that each argument supplied for a custom command can be autocompleted using <tab> in the same way that the first argument supplied can be.

### Configuration

| key                | value                                                            |
| ------------------ | ---------------------------------------------------------------- |
| version            | 0.99.1                                                           |
| major              | 0                                                                |
| minor              | 99                                                               |
| patch              | 1                                                                |
| branch             | makepkg                                                          |
| commit_hash        | 8812072f065be283df3e9a81f3945d0364f2f247                         |
| build_os           | linux-x86_64                                                     |
| build_target       | x86_64-unknown-linux-gnu                                         |
| rust_version       | rustc 1.82.0 (f6e511eec 2024-10-15) (Arch Linux rust 1:1.82.0-2) |
| cargo_version      | cargo 1.82.0 (8f40fc59f 2024-08-21)                              |
| build_time         | 2024-10-21 09:58:41 +00:00                                       |
| build_rust_channel | release                                                          |
| allocator          | mimalloc                                                         |
| features           | default, sqlite, trash                                           |
| installed_plugins  |                                                                  |

**Comment by RobbingDaHood** [comment]
Tried adding this patch to test for the above, but it passes fine. 

<details>
  <summary>Patch of added tests </summary>

``` 

Index: crates/nu-cli/tests/completions/mod.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/nu-cli/tests/completions/mod.rs b/crates/nu-cli/tests/completions/mod.rs
--- a/crates/nu-cli/tests/completions/mod.rs	(revision HEAD)
+++ b/crates/nu-cli/tests/completions/mod.rs	(revision Staged)
@@ -329,6 +329,72 @@
     // Match the results
     match_suggestions(&expected_paths, &suggestions);
 
+    // Test completions for the current folder even with parts before the autocomplet
+    let target_dir = format!("cp somefile.txt {dir_str}{MAIN_SEPARATOR}");
+    let suggestions = completer.complete(&target_dir, target_dir.len());
+
+    // Create the expected values
+    let expected_paths: Vec<String> = vec![
+        folder(dir.join("another")),
+        file(dir.join("custom_completion.nu")),
+        folder(dir.join("directory_completion")),
+        file(dir.join("nushell")),
+        folder(dir.join("test_a")),
+        folder(dir.join("test_b")),
+        file(dir.join(".hidden_file")),
+        folder(dir.join(".hidden_folder")),
+    ];
+
+    #[cfg(windows)]
+    {
+        let separator = '/';
+        let target_dir = format!("cp {dir_str}{separator}");
+        let slash_suggestions = completer.complete(&target_dir, target_dir.len());
+
+        let expected_slash_paths: Vec<String> = expected_paths
+            .iter()
+            .map(|s| s.replace('\\', "/"))
+            .collect();
+
+        match_suggestions(&expected_slash_paths, &slash_suggestions);
+    }
+
+    // Match the results
+    match_suggestions(&expected_paths, &suggestions);
+
+    // Test completions for the current folder even with parts before the autocomplet and custom commands
+    let target_dir = format!("my-command somefile.txt {dir_str}{MAIN_SEPARATOR}");
+    let suggestions = completer.complete(&target_dir, target_dir.len());
+
+    // Create the expected values
+    let expected_paths: Vec<String> = vec![
+        folder(dir.join("another")),
+        file(dir.join("custom_completion.nu")),
+        folder(dir.join("directory_completion")),
+        file(dir.join("nushell")),
+        folder(dir.join("test_a")),
+        folder(dir.join("test_b")),
+        file(dir.join(".hidden_file")),
+        folder(dir.join(".hidden_folder")),
+    ];
+
+    #[cfg(windows)]
+    {
+        let separator = '/';
+        let target_dir = format!("cp {dir_str}{separator}");
+        let slash_suggestions = completer.complete(&target_dir, target_dir.len());
+
+        let expected_slash_paths: Vec<String> = expected_paths
+            .iter()
+            .map(|s| s.replace('\\', "/"))
+            .collect();
+
+        match_suggestions(&expected_slash_paths, &slash_suggestions);
+    }
+
+    // Match the results
+    match_suggestions(&expected_paths, &suggestions);
+
     // Test completions for a file
     let target_dir = format!("cp {}", folder(dir.join("another")));
     let suggestions = completer.complete(&target_dir, target_dir.len());

``` 

</details>

So the last addition seems to add the test that should fail. 

I did also try testing the above locally and it still fails. 

BTW, the example in the description is missing an `s` in the last `args`:  
```
def list [...args] {
    $args
    | each {
        open $args
    }
}
```

**Comment by RobbingDaHood** [comment]
Can I be assigned this bug? 

I have a proposed fix, but I am not sure if that is the wanted solution: If it is then I will make a branch and add some tests. 

```
Index: crates/nu-parser/src/parser.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/nu-parser/src/parser.rs b/crates/nu-parser/src/parser.rs
--- a/crates/nu-parser/src/parser.rs	(revision Staged)
+++ b/crates/nu-parser/src/parser.rs	(date 1732606537864)
@@ -3712,7 +3712,7 @@
                                 args.push(Arg::RestPositional(PositionalArg {
                                     desc: String::new(),
                                     name,
-                                    shape: SyntaxShape::Any,
+                                    shape: SyntaxShape::OneOf(vec![SyntaxShape::GlobPattern, SyntaxShape::Any]),
                                     var_id: Some(var_id),
                                     default_value: None,
                                 }));
```  

The file completion only works if the previous parameter were a glob too: So that is why it first breaks at the first parameter.

**Comment by dyskos** [comment]
Not sure if this is related, but I also found that Nushell refuses to autocomplete commands that are not the precise case, despite case-sensitivity being off.

I found out after installing Racket on Windows, and trying to autocomplete Racket from the command-line as `rack<tab>`. Similarly to the original problem, all I receive is `NO RECORDS FOUND`. The executable is named `Racket.exe`. If I type `Rack<tab>` it will autocomplete to `Racket.exe`. And if I type `racket<enter>`, it will still run the executable correctly, it will just refuse to autocomplete.

If this is a fundamentally separate issue though, I can open another issue to track this.
